File: rust-executor/src/services/secrets.rs
use anyhow::{anyhow, Result};
use base64::{engine::general_purpose::STANDARD, Engine};
use ring::aead::{Aad, LessSafeKey, Nonce, UnboundKey, AES_256_GCM};
use serde_json::{Value, Map};
use sqlx::{Pool, Postgres, Row};
use std::collections::HashMap;
use std::env;
use sha2::{Sha256, Digest};
use tracing::{error, info, warn};

const IV_LENGTH: usize = 12; // 96 bits for GCM

/// Get the master encryption key from environment
fn get_master_key() -> Result<Vec<u8>> {
    let key_str = env::var("SECRETS_ENCRYPTION_KEY")
        .map_err(|_| anyhow!("SECRETS_ENCRYPTION_KEY environment variable is not set"))?;

    // Hash the key to get consistent 256-bit key (same as Python implementation)
    let mut hasher = Sha256::new();
    hasher.update(key_str.as_bytes());
    Ok(hasher.finalize().to_vec())
}

/// Decrypt a secret value
///
/// The encrypted value is base64-encoded and contains: IV (12 bytes) + ciphertext + tag
fn decrypt_secret(encrypted_value: &str) -> Result<String> {
    let master_key = get_master_key()?;

    // Decode from base64
    let combined = STANDARD.decode(encrypted_value)
        .map_err(|e| anyhow!("Failed to decode base64: {}", e))?;

    if combined.len() < IV_LENGTH {
        return Err(anyhow!("Encrypted value too short"));
    }

    // Extract IV and ciphertext+tag
    let (iv, ciphertext_and_tag) = combined.split_at(IV_LENGTH);

    // Create unbound key and nonce
    let unbound_key = UnboundKey::new(&AES_256_GCM, &master_key)
        .map_err(|e| anyhow!("Failed to create key: {:?}", e))?;
    let key = LessSafeKey::new(unbound_key);
    let nonce = Nonce::try_assume_unique_for_key(iv)
        .map_err(|_| anyhow!("Invalid nonce length"))?;

    // Decrypt
    let mut in_out = ciphertext_and_tag.to_vec();
    key.open_in_place(nonce, Aad::empty(), &mut in_out)
        .map_err(|e| anyhow!("Decryption failed: {:?}", e))?;

    // Remove the tag (last 16 bytes) from the result
    let tag_len = AES_256_GCM.tag_len();
    let plaintext_len = in_out.len().saturating_sub(tag_len);
    in_out.truncate(plaintext_len);

    String::from_utf8(in_out)
        .map_err(|e| anyhow!("Failed to convert to UTF-8: {}", e))
}

/// Load and decrypt all secrets for an organization
pub async fn load_org_secrets(pool: &Pool<Postgres>, org_id: &str) -> Result<HashMap<String, String>> {
    let mut secrets = HashMap::new();

    let rows = sqlx::query(
        r#"
        SELECT name, encrypted_value
        FROM org_secrets
        WHERE org_id = $1
        "#
    )
    .bind(org_id)
    .fetch_all(pool)
    .await?;

    for row in rows {
        let name: String = row.get("name");
        let encrypted_value: String = row.get("encrypted_value");

        match decrypt_secret(&encrypted_value) {
            Ok(decrypted_value) => {
                secrets.insert(name.clone(), decrypted_value);
                info!("Loaded secret: {}", name);
            }
            Err(e) => {
                error!("Failed to decrypt secret {}: {}", name, e);
                // Continue loading other secrets even if one fails
            }
        }
    }

    Ok(secrets)
}

/// Inject secrets into workflow parameters
///
/// Supports two modes:
/// 1. Placeholder substitution: Replace ${SECRET_NAME} with actual values
/// 2. Direct injection: Add secrets as top-level parameters
pub fn inject_secrets_into_params(
    params: Value,
    secrets: &HashMap<String, String>,
    substitute_placeholders: bool,
) -> Value {
    let mut result = params.clone();

    // Mode 1: Substitute placeholders in existing parameters
    if substitute_placeholders {
        result = substitute_placeholders_recursive(result, secrets);
    }

    // Mode 2: Add secrets as top-level parameters
    // This allows workflows to access secrets directly as variables
    if let Value::Object(ref mut map) = result {
        for (name, value) in secrets {
            // Add secret with original name
            if !map.contains_key(name) {
                map.insert(name.clone(), Value::String(value.clone()));
            }

            // Also add lowercase version for compatibility
            // E.g., APOLLO_API_KEY -> apollo_api_key
            let lowercase_name = name.to_lowercase();
            if !map.contains_key(&lowercase_name) {
                map.insert(lowercase_name, Value::String(value.clone()));
            }
        }
    }

    result
}

/// Recursively substitute ${SECRET_NAME} placeholders in JSON values
fn substitute_placeholders_recursive(value: Value, secrets: &HashMap<String, String>) -> Value {
    match value {
        Value::String(s) => {
            let mut result = s.clone();
            for (name, secret_value) in secrets {
                let placeholder = format!("${{{}}}", name);
                result = result.replace(&placeholder, secret_value);
            }
            Value::String(result)
        }
        Value::Object(map) => {
            let mut new_map = Map::new();
            for (k, v) in map {
                new_map.insert(k, substitute_placeholders_recursive(v, secrets));
            }
            Value::Object(new_map)
        }
        Value::Array(arr) => {
            Value::Array(arr.into_iter().map(|v| substitute_placeholders_recursive(v, secrets)).collect())
        }
        _ => value,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn test_substitute_placeholders() {
        let mut secrets = HashMap::new();
        secrets.insert("GITHUB_TOKEN".to_string(), "ghp_secret123".to_string());
        secrets.insert("API_KEY".to_string(), "key456".to_string());

        let input = json!({
            "token": "${GITHUB_TOKEN}",
            "api_key": "${API_KEY}",
            "nested": {
                "value": "prefix ${GITHUB_TOKEN} suffix"
            },
            "unchanged": "no placeholder here"
        });

        let result = substitute_placeholders_recursive(input, &secrets);

        assert_eq!(result["token"], "ghp_secret123");
        assert_eq!(result["api_key"], "key456");
        assert_eq!(result["nested"]["value"], "prefix ghp_secret123 suffix");
        assert_eq!(result["unchanged"], "no placeholder here");
    }

    #[test]
    fn test_inject_secrets() {
        let mut secrets = HashMap::new();
        secrets.insert("GITHUB_TOKEN".to_string(), "ghp_secret123".to_string());

        let input = json!({
            "existing_param": "value"
        });

        let result = inject_secrets_into_params(input, &secrets, false);

        // Should add both GITHUB_TOKEN and github_token
        assert_eq!(result["existing_param"], "value");
        assert_eq!(result["GITHUB_TOKEN"], "ghp_secret123");
        assert_eq!(result["github_token"], "ghp_secret123");
    }
}

File: modal_apps/lib/secrets.py
"""
Secrets management utilities for workflow execution
Handles decryption and injection of org-level secrets
"""
import base64
import hashlib
import logging
import os
from typing import Dict, Optional

from cryptography.hazmat.primitives.ciphers.aead import AESGCM

logger = logging.getLogger(__name__)

# Constants
ALGORITHM = "AES-GCM"
IV_LENGTH = 12  # 96 bits recommended for GCM


def get_master_key() -> bytes:
    """Get the master encryption key from environment"""
    key_str = os.environ.get("SECRETS_ENCRYPTION_KEY")
    if not key_str:
        raise ValueError("SECRETS_ENCRYPTION_KEY environment variable is not set")

    # Hash the key to get consistent 256-bit key
    key_hash = hashlib.sha256(key_str.encode()).digest()
    return key_hash


def decrypt_secret(encrypted_value: str) -> str:
    """
    Decrypt a secret value

    Args:
        encrypted_value: Base64-encoded encrypted value with IV prepended

    Returns:
        Decrypted plaintext secret
    """
    try:
        master_key = get_master_key()

        # Decode from base64
        combined = base64.b64decode(encrypted_value)

        # Extract IV and ciphertext
        iv = combined[:IV_LENGTH]
        ciphertext = combined[IV_LENGTH:]

        # Initialize AESGCM cipher
        aesgcm = AESGCM(master_key)

        # Decrypt
        plaintext = aesgcm.decrypt(iv, ciphertext, None)

        return plaintext.decode("utf-8")
    except Exception as e:
        logger.error(f"Failed to decrypt secret: {e}")
        raise ValueError(f"Failed to decrypt secret: {e}")


def load_org_secrets(conn, org_id: str) -> Dict[str, str]:
    """
    Load and decrypt all secrets for an organization

    Args:
        conn: Database connection
        org_id: Organization ID

    Returns:
        Dictionary of secret_name -> decrypted_value
    """
    secrets = {}

    try:
        with conn.cursor() as cur:
            cur.execute(
                """
                SELECT name, encrypted_value
                FROM org_secrets
                WHERE org_id = %s
                """,
                (org_id,)
            )

            rows = cur.fetchall()

            for row in rows:
                name, encrypted_value = row
                try:
                    decrypted_value = decrypt_secret(encrypted_value)
                    secrets[name] = decrypted_value
                    logger.info(f"Loaded secret: {name}")
                except Exception as e:
                    logger.error(f"Failed to decrypt secret {name}: {e}")
                    # Continue loading other secrets even if one fails

    except Exception as e:
        logger.error(f"Failed to load org secrets: {e}")
        # Return empty dict if loading fails - don't block workflow execution

    return secrets


def inject_secrets_into_params(
    params: Dict,
    secrets: Dict[str, str],
    substitute_placeholders: bool = True
) -> Dict:
    """
    Inject secrets into workflow parameters

    Supports two modes:
    1. Placeholder substitution: Replace ${SECRET_NAME} with actual values
    2. Direct injection: Add secrets as top-level parameters

    Args:
        params: Workflow execution parameters
        secrets: Dictionary of secret_name -> value
        substitute_placeholders: If True, replace ${SECRET_NAME} in string values

    Returns:
        Updated parameters with secrets injected
    """
    result = params.copy()

    # Mode 1: Substitute placeholders in existing parameters
    if substitute_placeholders:
        result = _substitute_placeholders(result, secrets)

    # Mode 2: Add secrets as top-level parameters
    # This allows workflows to access secrets directly as variables
    # Support both UPPERCASE and lowercase_with_underscores naming conventions
    for name, value in secrets.items():
        # Add secret with original name
        if name not in result:
            result[name] = value

        # Also add lowercase version for compatibility
        # E.g., APOLLO_API_KEY -> apollo_api_key
        lowercase_name = name.lower()
        if lowercase_name not in result:
            result[lowercase_name] = value

    return result


def _substitute_placeholders(obj, secrets: Dict[str, str]):
    """
    Recursively substitute ${SECRET_NAME} placeholders in strings
    """
    if isinstance(obj, str):
        # Replace ${SECRET_NAME} with actual value
        for name, value in secrets.items():
            placeholder = f"${{{name}}}"
            if placeholder in obj:
                obj = obj.replace(placeholder, value)
        return obj
    elif isinstance(obj, dict):
        return {k: _substitute_placeholders(v, secrets) for k, v in obj.items()}
    elif isinstance(obj, list):
        return [_substitute_placeholders(item, secrets) for item in obj]
    else:
        return obj
